### PR TITLE
fix: fix priority scheduler preemption and prioritization issues

### PR DESCRIPTION
**Pull request by hamidzr**
## Description

Fixes two separate issues when priority between tasks/jobs are equal in the priority scheduler and it has fall back on its secondary sorting logic. Assuming that we want the scheduler to prioritize the oldest user-submitted ENTbCS (aka jobs)

1. the sorting was based on the Allocation actor's registered time which is unstable, changed during different cluster events, and would easily go out of sync with "user submission time"
2. the scheduler picked the oldest user-submitted job to preempt when needed instead of the most recent submitted one

## Commentary (optional)

issue discussion, demo, and reproduction steps: https://determined-community.slack.com/archives/CV3MVAQJU/p1639766427055600

## Checklist

- [x] User-facing API changes need the "User-facing API Change" label.
- [x] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [x] Licenses should be included for new code which was copied and/or modified from any external code.

[DET-1234]: https://determinedai.atlassian.net/browse/DET-1234?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: master/internal/db/postgres_tasks_intg_test.go [hunk 1]
```diff
@@ -10,12 +10,13 @@ import (
 	"time"
 
 	"github.com/davecgh/go-spew/spew"
```
Comment (hamidzr) [inline master/internal/db/postgres_tasks_intg_test.go:12]: revert this

File: master/internal/checkpoint_gc.go [hunk 2]
```diff
@@ -14,6 +14,7 @@ import (
 	"github.com/determined-ai/determined/master/pkg/tasks"
 )
 
+// TODO these are job related tasks as well.
 type checkpointGCTask struct {
```
Comment (hamidzr) [inline master/internal/checkpoint_gc.go:18]: how do these get scheduled. what priority/weight do they take on? if I"m not mistaken they're 0 slot so they get scheduled separately

**Comment by azhou-determined** [comment]
👍 tested this fix locally by submitting 4 experiments with priority 20, verified they are being run in the order of submission.

**Comment by azhou-determined** [comment]
also modified some of the unit tests in `priority_test.go` here: https://github.com/azhou-determined/determined/tree/unit-tests-priority-schedule-job-time
can merge if it looks good

File: master/internal/resourcemanagers/priority.go [hunk 3]
```diff
@@ -185,7 +185,8 @@ func trySchedulingTaskViaPreemption(
 	log.Debugf("trying to schedule task %s by preempting other tasks", allocationRequest.Name)
 
 	for priority := model.MaxUserSchedulingPriority; priority > allocationPriority; priority-- {
-		for _, preemptionCandidate := range priorityToScheduledTaskMap[priority] {
+		for i := len(priorityToScheduledTaskMap[priority]) - 1; i >= 0; i-- {
```
Comment (stoksc) [inline master/internal/resourcemanagers/priority.go:188]: idea here is we should preempt newest jobs first when trying to make room for high priority jobs?

File: master/internal/resourcemanagers/priority.go [hunk 4]
```diff
@@ -244,8 +245,8 @@ func sortTasksByPriorityAndTimestamp(
 	priorityToPendingTasksMap := make(map[int][]*sproto.AllocateRequest)
 	priorityToScheduledTaskMap := make(map[int][]*sproto.AllocateRequest)
 
-	for it := taskList.iterator(); it.next(); {
-		req := it.value()
+	reqs := sortTasks(taskList, groups, false)
```
Comment (stoksc) [inline master/internal/resourcemanagers/priority.go:248]: i dont think we need this change if we also change the comparator passed to treeset.NewWith, so that the iterator is still in the correct order? if so, why this change? after a quick glance, that looks more correct for elsewhere the iterator is used (where it matters, and order is important).

File: master/internal/resourcemanagers/priority.go (hunk 3 above)
Comment (hamidzr) [reply master/internal/resourcemanagers/priority.go:188]: yes an extension of "we prioritize jobs that have been submitted the longest" which I think was also the original intent?

File: master/internal/resourcemanagers/priority.go (hunk 4 above)
Comment (hamidzr) [reply master/internal/resourcemanagers/priority.go:248]: yeah we shouldn't need to but I wanted to make sure there is less chance of deviation on how the jobs are sorted but I can take it out for this particular fix.

Comment (hamidzr) [reply master/internal/resourcemanagers/priority.go:248]: we could pass in the filter to  sortTasks too if that's a concern.

File: master/internal/experiment.go [hunk 5]
```diff
@@ -444,8 +444,8 @@ func (e *experiment) processOperations(
 			state := trialSearcherState{Create: op, Complete: true}
 			e.TrialSearcherState[op.RequestID] = state
 			ctx.ActorOf(op.RequestID, newTrial(
-				trialTaskID(e.ID, op.RequestID), e.JobID, e.ID, e.State, state, e.rm, e.trialLogger, e.db,
-				config, checkpoint, e.taskSpec,
+				trialTaskID(e.ID, op.RequestID), e.JobID, e.StartTime, e.ID, e.State, state, e.rm,
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/experiment.go:447]: what kind of arguments are you thinking it would encompass? would the trial keep it in memory that way too?

**Comment by hamidzr** [comment]
> also modified some of the unit tests in `priority_test.go` here: https://github.com/azhou-determined/determined/tree/unit-tests-priority-schedule-job-time can merge if it looks good

thank you. I also cherry-picked the tests here

File: master/internal/resourcemanagers/scheduler_test.go [hunk 6]
```diff
@@ -435,6 +438,18 @@ func assertEqualToAllocate(
 		"actual tasks and expected tasks must have the same length")
 }
 
+func assertEqualToAllocateOrdered(
```
[the remark this answers is not in the document]
Comment (hamidzr) [reply master/internal/resourcemanagers/scheduler_test.go:441]: I didn't write this but It seems to go off of `assertEqualToAllocate`. any name suggestions? these are the other ones we have. could probably be added as a parameter to `assertEqualToAllocate` too.

```
master/internal/resourcemanagers/scheduler_test.go:424:func assertEqualToAllocate(
master/internal/resourcemanagers/scheduler_test.go:441:func assertEqualToAllocateOrdered(
master/internal/resourcemanagers/scheduler_test.go:453:func assertEqualToRelease(
master/internal/resourcemanagers/provisioner/scale_decider_test.go:13:func assertEqualInstancesMarked(t
master/pkg/schemas/merge.go:46:func assertTypeMatch(obj
master/test/integration/api/api_trials_test.go:605:func assertStringContains(t

```

File: master/internal/experiment.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (stoksc) [reply master/internal/experiment.go:447]: this was just me pointing out this function has 12 params. let's handle it later.

File: master/internal/resourcemanagers/priority.go (hunk 3 above)
Comment (stoksc) [reply master/internal/resourcemanagers/priority.go:188]: makes sense.

File: master/internal/sproto/task.go [hunk 7]
```diff
@@ -15,13 +15,15 @@ type (
...
-		TaskID       model.TaskID
-		JobID        *model.JobID
-		State        job.SchedulingState
-		Name         string
-		TaskActor    *actor.Ref
-		Group        *actor.Ref
+		AllocationID      model.AllocationID
+		TaskID            model.TaskID
+		JobID             *model.JobID
+		JobSubmissionTime *time.Time // TODO or a job info struct
```
Comment (stoksc) [inline master/internal/sproto/task.go:21]: i think we should either cleanup these todos or make a ticket for them

File: docs/release-notes/3428-fix-priority-scheduler-timelogic.txt [hunk 8]
```diff
@@ -0,0 +1,6 @@
+:orphan:
+
+**Bug Fixes**
+
+-  Master: Fixes two separate issues when priority between tasks/jobs is equal in priority scheduler
+   and it falls back on its secondary sorting logic.
```
Comment (stoksc) [inline docs/release-notes/3428-fix-priority-scheduler-timelogic.txt:6]:
```suggestion
-  Master: Fix a bug in the priority scheduler where jobs with equal priority would be scheduled or preempted in an order not correctly respecting job submission time.
```

File: master/internal/resourcemanagers/priority.go (hunk 3 above)
Comment (stoksc) [inline master/internal/resourcemanagers/priority.go:188]: nit golang, why no slice.Reverse...

**Review by stoksc — APPROVED** [review]
this by itself is fine but i still think we might should change `master/internal/resourcemanagers/task_list.go`'s `taskComparator` to use `compareByRegisteredTime` rather than also still using registered time.

File: master/internal/sproto/task.go (hunk 7 above)
Comment (hamidzr) [reply master/internal/sproto/task.go:21]: since I didn't have clear stories for addressing them I just cleaned them up

File: master/internal/resourcemanagers/priority.go (hunk 3 above)
Comment (hamidzr) [reply master/internal/resourcemanagers/priority.go:188]: sort.Reverse? I didn't know about it. and that seems to not modify the array in place which we can't have.

**Comment by hamidzr** [comment]
> this by itself is fine but i still think we might should change `master/internal/resourcemanagers/task_list.go`'s `taskComparator` to use `compareByRegisteredTime` rather than also still using registered time.

makes sense. I'll update that after this lands

File: master/internal/resourcemanagers/priority.go (hunk 3 above)
Comment (stoksc) [reply master/internal/resourcemanagers/priority.go:188]: Oh I mean nit the go language lol
